keps/sig-network/1880-multiple-service-cidrs/README.md

```diff
@@ -747,53 +737,87 @@ This section must be completed when targeting beta to a release.
 
 ###### How can a rollout or rollback fail? Can it impact already running workloads?
 
-<!--
-Try to be as paranoid as possible - e.g., what if some components will restart
-mid-rollout?
+The objects consumed by the cluster components are the Services, and ClusterIP is a field of these objects.
+Running workloads can not be impacted since the existing Services objects are never mutated.
 
-Be sure to consider highly-available clusters, where, for example,
-feature flags will be enabled on some API servers and not others during the
-rollout. Similarly, consider large clusters and how enablement/disablement
-will rollout across nodes.
--->
+A rollout failure can impact the new apiserver to not be able to start or the clients not being able to create new Services.
 
 ###### What specific metrics should inform a rollback?
 
-<!--
-What signals should users be paying attention to when the feature is young
-that might indicate a serious problem?
--->
+The feature impact the apiserver bootstrap process, specially about the kubernetes.default IP address assignment,
+in case the apiserver is not able to start after enabling the feature, it is a strong indicated that a rollback is required.
+
+Another metrics that can indicate a rollback are `clusterip_allocator.allocation_errors_total`, `clusterip_repair.ip_errors_total` or `clusterip_repair.reconcile_errors_total`, definitions can be found on
+- IP allocator pkg/registry/core/service/ipallocator/metrics.go
+- IP repair loop pkg/registry/core/service/ipallocator/controller/metrics.go
 
 ###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
 
-<!--
-Describe manual testing that was done and the outcomes.
-Longer term, we may want to require automated upgrade/rollback tests, but we
-are missing a bunch of machinery and tooling and can't do that now.
--->
+There are integration tests these behaviors:
+
+test/integration/servicecidr/allocator_test.go TestSkewedAllocators
+1. start one apiserver with the feature gate enabled and other with the feature gate disabled
+2. create 5 Services against each apiserver
+3. the Services created against the apiserver with the bitmap will not allocate the IPAddresses automatically,
```

Comment:
To clarify: These will be allocated in the bitmap but not as IPAddress objects, but the OTHER apiserver will allocate them as IPAddress objects in the repair loop, right?

Is it possible to have both apiservers allocate the same IP ?  I think it is, right?
  - the old API allocates an IP and writes the bitmap
  - the new API allocates the same IP and writes the IPAddress
  - the repair loop catches up and finds a service (from old API) with an IP that doesn't point back to it
  
Am I forgetting some place where we prevent that?  What can we do about it?

What if the new allocator also wrote the bitmap (when it is allocating from the "kubernetes" range(s))?

 Do we need to do a release that always writes IPAddress objects but does't allow multiple CIDRs?  That way the NEXT release will always collide on IPAddress?

Reply:
Similar, but with a single APIserver:

* Boot with feature disabled
* Create service with IP X 
* Enable feature
* Auto-create the flags-based range
* Create another service with same IP
* Repair loop catches up

Reply:
Writing down the conversations we had yesterday and some thoughts.
Some details of existing implementation
- The apiserver will not be ready until one execution of the repair loop is succesful https://github.com/kubernetes/kubernetes/blob/c3689b9f8b8c1a1ac2bbc4e3b3cda28d83849ec2/pkg/registry/core/rest/storage_core.go#L432-L470
- The single API server scenarios (last comment) is tested with an integration test, see below ` TestMigrateService` 

> the Services created against the apiserver with the bitmap will not allocate the IPAddresses automatically,

The repair loop in the new allocator watches Services events, at the moment that a new Service is created in the old apiserver it will reconcile and create the corresponding IPAddress object, this reduces the time window where the race is possible.

Problem statement in a scenario with 2 apiservers;
- apiserver A has feature enabled and apiserverB disabled and both are ready
- t1 is the time necessary by A to get a Service creation event and reconcile the corresponding IPAddress
- t2 is the time necessary by B to reconcile the Service IPs created by A, this right now is triggered periodically but can be implemented as triggered by events too

The problem is when there are two Services created at the same time requesting the same IP.
If the IP is randomly allocated we are in a birthday paradox like problem, what are the chances two different allocators choose the same value? it will depend on the serviceCIDR size ...

If the IP is statically allocated, this is the worst scenario, though not seem very realistic, but the consequences are bad, if two different Services ask for the same IP:
- in current version, one of these requests will fail with skewed apiservers
- with current scenario, if one of each requests go to different apiservers, it will depend on the window of time defined by the max(t1,t2) and may create two Services with the same IP ... the repair IP will start generating events but this will depend on the admin to realize of the problem

The simplest solution is to move the old repair loop to receive the Service create objects event to trigger a repair loop, this will reduce significantly the time window and we may rely on statistics to justify this decision (there are a lot of ifs that has to happen in a small window of time and use documentation to mitigate the risks.

I was thinking in providing a tool to run against the cluster to detect this and also to offer the option of delete the old bitmaps stored ... we have done something similar in the past with the storage version migration https://github.com/kubernetes-sigs/kube-storage-version-migrator, (interestingly this is moving in tree in 1.30 https://kubernetes.io/docs/tasks/manage-kubernetes-objects/storage-version-migration/ )

If we want to get a zero chance of this to happen, we need to sync both apiservers, this means we need some sort of synchronization between them, and it seems the opaque bitmap for the old allocator is the best alternative.
@thockin suggested an additional feature gate that will allow us to do a 2 phases rollout and it can work to solve the upgrade/skew problem, however, it does not solve the long term problem, because one of the main uses cases of this new feature is to be able to shrink or grow the default service cidr without having to restart the apiservers, and this will imply to resize the bitmap allocator, that will break the skewed servers

Reply:
Capturing discussion, the idea:

* Add a new gate, e.g. "ServiceCIDRLegacyBitmap", stage Deprecated, starts as beta default=on (captures the existing behavior)
* Add logic to new-path allocator to write BOTH the `IPAddress` resources *and* the bitmap when allocating from the flag-derived range (bitmap first, I think)
* Roll this all out as beta in 1.31
* In case of version skew described above, both old and new will use the bitmap so no collision
* In 1.32 change gate ServiceCIDRLegacyBitmap to default=off
* When we move the real gate to GA, add code to delete the the bitmap entirely
* Next release we remove that cleanup code

@aojea - agree or did I get some detail wrong?

Reply:
> * In 1.32 change gate ServiceCIDRLegacyBitmap to default=off

Do you envision default=off and lock=true, or users can still turn this back on. In which case I believe you fallback to the skew scenario, and you write both, did I get that correct?

Reply:
Turning it ON shouldn't hurt and it's nice to be able to fall back if we did miss something. So I don't think we would lock it, at least not in that first release.  In fact, maybe we don't need the gate in this release at all. We really don't want people to turn it off, I think. 

So 31 double writes always. 32 has double write disabled but can be enabled. 33 has double right disabled. And 34 removes it?  That would correspond with beta and 31 ga and 32 remove bitmap in 33 and final cleanup and 34?

Reply:
ok, I have the implementation and a very exhaustive test of the problem https://github.com/kubernetes/kubernetes/pull/122047/commits/b29cbfaf1664acf84cb08cf1f9d8bc6cd757ebb1 so I'm confident on the solution.

Regarding the rollout of the feature, I was thinking in using the negative


| Version | MultiCIDRServiceAllocator | DisableAllocatorDualWrite |
|----------|----------|----------|
| 1.31    | Beta off   | Alpha off   |
| 1.32   | GA on  (locked) | Beta off   |
| 1.33    | GA on  (there are no bitmaps running) | GA on (also delete old bitmap)|
| 1.34    | remove feature gate | remove feature gate  |

Reply:
I am worried about jumping from Beta,off -> GA,on,locked.  I think we want at least 1 cycle of Beta,on - disagree?

Reply:
there was a thread in slack about it, I was going initially with 2 beta cycles, but that thread seemed to encourage going GA https://kubernetes.slack.com/archives/C0EG7JC6T/p1718289997541569?thread_ts=1718249135.862569&cid=C0EG7JC6T

To be honest, at this point I don't think we expect much changes in the API, so maybe is better to go GA without the lock to avoid the beta APIs to be lingering on the clusters, the effect will be the same

Reply:
We don't have to pin it down here/yet.  My fear is that the move from "off" to "on" is the riskiest moment, and if "on" means "GA", we have no recourse.

Reply:
agree, if the main concern are the API objects, can we move the API to v1 and leave the feature gate in beta?

Reply:
I don't think that helps, and is probably more confusing overall.

I think we need a cycle of beta+on -- that could be this cycle, or it could be 32, but I think it has to happen

Reply:
ok, let me push the latest update, 1 beta off now, 1 beta on next